_posts/2020-10-27-solidity-0.8.x-preview.md

```diff
@@ -0,0 +1,192 @@
+   corresponds to the ABI-encoding of a function call to a function with the signature ``Error(string)``.
+ - critical errors revert with error data that corresponds to the ABI-encoding of a function call
+   to a function with the signature ``Panic(uint256)``.
+
+Because of that, we also use the term "panic" or "panic error" for such critical errors.
+
+To detect the two situations, you can take a look at the first four bytes of the return data in case
+of a failure. If it is ``0x4e487b71``, then it is a panic error, if it is ``0x08c379a0``
+or if the message is empty, it is a "regular" error.
+
+Panic errors use specific error codes to distinguish certain situations in which the panic error is triggered.
+This is the current list of error codes:
+
+#. 0x01: If you call ``assert`` with an argument that evaluates to false.
+#. 0x11: If an arithmetic operation results in underflow or overflow outside of an ``unchecked { ... }`` block.
+#. 0x12; If you divide or modulo by zero (e.g. ``5 / 0`` or ``23 % 0``).
+#. 0x21: If you convert a value that is too big or negative into an enum type.
+#. 0x31: If you call ``.pop()`` on an empty array.
+#. 0x32: If you access an array, ``bytesN`` or an array slice at an out-of-bounds or negative index (i.e. ``x[i]`` where ``i >= x.length`` or ``i < 0``).
+#. 0x41: If you allocate too much memory or create an array that is too large.
+#. 0x51: If you call a zero-initialized variable of internal function type.
+
+This list of codes can be extended in the future.
+
+#### Actual Checked Arithmetic
+
+By default, all arithmetic operations will perform overflow and underflow checks
+(Solidity already had division by zero checks). In case of an underflow or
+overflow, a ``Panic(0x11)`` error will be thrown and the call will revert.
+
+The following code, for example, will trigger such an error:
+
+```solidity
+contract C {
+    function f() public pure {
+        uint x = 0;
+        x--;
+    }
+}
+```
+
+The checks are based on the actual type of the variable, so even if the result would fit
+an EVM word of 256 bits, if your type is ``uint8`` and if the result is larger than ``255``,
+it will trigger a ``Panic``.
+
+We did not introduce checks for explicit type conversions from larger to smaller
+types, because we think that such checks might be unexpected. We would love to get
+your input on that matter!
+
+The following operations now have checks that they did not have before:
+ - addition (``+``), subtraction (``-``), multiplication (``*``)
+ - increment and decrement (``++`` / ``--``)
+ - unary negation (``-``)
+ - exponentiation (``**``)
+ - division (see below) (``/``)
+
+There are some edge cases you might not expect.
+For example, unary negation is only available for signed types.
+The problem is that in two's complement representation, there is one more negative number for each bit width
+than there are positive numbers. This means that the folloing code will trigger a Panic:
+
+```solidity
+int8 x = -128;
```

Comment:
All the other examples (except these two) in the post use a complete contract/function snippet.